### DISCUSSION TITLE
Error while deploying

### DISCUSSION THREAD

**Discussion by iober** [dokku/dokku#5358] [Q&A]
I have dokku on a VPS hosted on hostinger and I'm trying to send a node API, but I'm having difficulties.
The same API is working locally and also on Heroku.
Where am I wrong?
I'm a beginner, so I apologize for any mistakes.

**Error Dokku:**
```
PS C:\_Projetos\pessoal\app-agenda\app-agenda-back> git push dokku main
Enumerating objects: 41, done.
Counting objects: 100% (41/41), done.
Delta compression using up to 12 threads
Compressing objects: 100% (40/40), done.
Writing objects: 100% (41/41), 45.23 KiB | 2.66 MiB/s, done.
Total 41 (delta 22), reused 0 (delta 0), pack-reused 0
remote: -----> Cleaning up...
remote: -----> Building agenda-api from herokuish
remote: -----> Adding BUILD_ENV to build environment...
remote:        BUILD_ENV added successfully
remote:        -----> Node.js app detected
remote: 
remote: -----> Creating runtime environment
remote: 
remote:        NPM_CONFIG_LOGLEVEL=error
remote:        NODE_VERBOSE=false
remote:        NODE_ENV=production
remote:        NODE_MODULES_CACHE=true
remote:
remote: -----> Installing binaries
remote:        engines.node (package.json):  16.17.0
remote:        engines.npm (package.json):   8.15.0
remote:
remote:        Resolving node version 16.17.0...
remote:        Downloading and installing node 16.17.0...
remote:
remote: -----> Build failed
remote:
remote:        We're sorry this build is failing! You can troubleshoot common issues here:
remote:        https://devcenter.heroku.com/articles/troubleshooting-node-deploys
remote:
remote:        If you're stuck, please submit a ticket so we can help:
remote:        https://help.heroku.com/
remote:
remote:        Love,
remote:        Heroku
remote:
remote:  !     Failure during app build
remote:  !     Removing invalid image tag dokku/agenda-api:latest
remote:  !     App build failed
To iober.com.br:agenda-api
 ! [remote rejected] main -> main (pre-receive hook declined)
error: failed to push some refs to '*****:agenda-api'
PS C:\_Projetos\pessoal\app-agenda\app-agenda-back> node -v
v16.17.0
PS C:\_Projetos\pessoal\app-agenda\app-agenda-back> npm -v
8.15.0
```

**Package.json:**
```
{
  "name": "api-app-agenda",
  "version": "1.0.0",
  "description": "api-app-agenda by Iober Henrique da Cruz",
  "main": "index.js",
  "dependencies": {
    "body-parser": "^1.20.0",
    "dotenv": "^16.0.0",
    "express": "^4.17.3",
    "moment": "^2.29.4",
    "nodemon": "^2.0.15",
    "pg": "^8.7.3",
    "sha1": "^1.1.1"
  },
  "scripts": {
    "start": "node index.js",
    "lint": "eslint --ext .js,.vue src",
    "test": "echo \"No test specified\" && exit 0"
  },
  "engines": {
    "node": "16.17.0",
    "npm": "8.15.0"
  },
  "devDependencies": {},
  "author": "Iober Henrique da Cruz",
  "license": "ISC"
}
```

**On Heroku, all very well:**
```
$ git push heroku master
Enumerating objects: 21, done.
Counting objects: 100% (21/21), done.
Delta compression using up to 12 threads
Compressing objects: 100% (12/12), done.
Writing objects: 100% (13/13), 39.40 KiB | 7.88 MiB/s, done.
Total 13 (delta 6), reused 0 (delta 0), pack-reused 0
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> Building on the Heroku-20 stack
remote: -----> Using buildpack: heroku/nodejs
remote: -----> Node.js app detected
remote:
remote: -----> Creating runtime environment
remote:
remote:        NPM_CONFIG_LOGLEVEL=error
remote:        NODE_VERBOSE=false
remote:        NODE_ENV=production
remote:        NODE_MODULES_CACHE=true
remote:
remote: -----> Installing binaries
remote:        engines.node (package.json):  16.17.0
remote:        engines.npm (package.json):   8.15.0
remote:
remote:        Resolving node version 16.17.0...
remote:        Downloading and installing node 16.17.0...
remote:        npm 8.15.0 already installed with node
remote:
remote: -----> Restoring cache
remote:        - npm cache
remote:
remote: -----> Installing dependencies
remote:        Installing node modules
remote:
remote:        added 204 packages, and audited 205 packages in 2s
remote:
remote:        24 packages are looking for funding
remote:          run `npm fund` for details
remote:
remote:        5 moderate severity vulnerabilities
remote:
remote:        To address all issues, run:
remote:          npm audit fix
remote:
remote:        Run `npm audit` for details.
remote:
remote: -----> Build
remote:
remote: -----> Caching build
remote:        - npm cache
remote:
remote: -----> Pruning devDependencies
remote:
remote:        up to date, audited 205 packages in 834ms
remote:
remote:        24 packages are looking for funding
remote:          run `npm fund` for details
remote:
remote:        5 moderate severity vulnerabilities
remote:
remote:        To address all issues, run:
remote:          npm audit fix
remote:
remote:        Run `npm audit` for details.
remote:
remote: -----> Build succeeded!
remote: -----> Discovering process types
remote:        Procfile declares types     -> (none)
remote:        Default types for buildpack -> web
remote:
remote: -----> Compressing...
remote:        Done: 34.6M
remote: -----> Launching...
remote:        Released v39
remote:        https://****.herokuapp.com/ deployed to Heroku
remote:
remote: Verifying deploy.... done.
To https://git.heroku.com/****.git
   df9f99d..28e8a5d  master -> master
```

**Comment by josegonzalez** [comment]
Can you post the output of `dokku report`?

**Reply by iober** [reply]
**Excuse me... is this it?**

[report.txt](https://github.com/dokku/dokku/files/9595308/report.txt)


root@app:/home# dokku report painelgov-api
> uname: Linux app.vps 5.4.0 #1 SMP Tue Jan 25 12:49:12 MSK 2022 x86_64 x86_64 x86_64 GNU/Linux
> memory: 
                     total        used        free      shared  buff/cache   available
       Mem:           1024         140         357           1         525         882
       Swap:           256           0         256
> docker version: 
       Client: Docker Engine - Community
        Version:           20.10.18
        API version:       1.41
        Go version:        go1.18.6
        Git commit:        b40c2f6
        Built:             Thu Sep  8 23:11:45 2022
        OS/Arch:           linux/amd64
        Context:           default
        Experimental:      true
       
       Server: Docker Engine - Community
        Engine:
         Version:          20.10.18
         API version:      1.41 (minimum version 1.12)
         Go version:       go1.18.6
         Git commit:       e42327a
         Built:            Thu Sep  8 23:09:37 2022
         OS/Arch:          linux/amd64
         Experimental:     false
        containerd:
         Version:          1.6.8
         GitCommit:        9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
        runc:
         Version:          1.1.4
         GitCommit:        v1.1.4-0-g5fd4c4d
        docker-init:
         Version:          0.19.0
         GitCommit:        de40ad0
> docker daemon info: 
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
       Client:
        Context:    default
        Debug Mode: true
        Plugins:
         app: Docker App (Docker Inc., v0.9.1-beta3)
         buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
         compose: Docker Compose (Docker Inc., v2.10.2)
         scan: Docker Scan (Docker Inc., v0.17.0)
       
       Server:
        Containers: 2
         Running: 0
         Paused: 0
         Stopped: 2
        Images: 7
        Server Version: 20.10.18
        Storage Driver: overlay2
         Backing Filesystem: extfs
         Supports d_type: true
         Native Overlay Diff: true
         userxattr: false
        Logging Driver: json-file
        Cgroup Driver: cgroupfs
        Cgroup Version: 1
        Plugins:
         Volume: local
         Network: bridge host ipvlan macvlan null overlay
         Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
        Swarm: inactive
        Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
        Default Runtime: runc
        Init Binary: docker-init
        containerd version: 9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
        runc version: v1.1.4-0-g5fd4c4d
        init version: de40ad0
        Security Options:
         seccomp
          Profile: default
        Kernel Version: 5.4.0
        Operating System: Ubuntu 20.04.4 LTS
        OSType: linux
        Architecture: x86_64
        CPUs: 1
        Total Memory: 1GiB
        Name: app.vps
        ID: INXR:G7AG:AMMJ:JZ4M:TAZK:6GK5:QCZZ:WBQH:UQ5Z:QLZ6:TYGN:CVVO
        Docker Root Dir: /var/lib/docker
        Debug Mode: false
        Registry: https://index.docker.io/v1/
        Labels:
        Experimental: false
        Insecure Registries:
         127.0.0.0/8
        Live Restore Enabled: false
       
> git version: git version 2.25.1
> sigil version: 0.9.0build+bc921b7
> herokuish version: 
       herokuish: 0.5.37
       buildpacks:
         heroku-buildpack-multi     v1.2.0
         heroku-buildpack-ruby      v244
         heroku-buildpack-nodejs    v198
         heroku-buildpack-clojure   v90
         heroku-buildpack-python    v214
         heroku-buildpack-java      v72
         heroku-buildpack-gradle    v38
         heroku-buildpack-scala     v94
         heroku-buildpack-play      v26
         heroku-buildpack-php       v223
         heroku-buildpack-go        v166
         heroku-buildpack-nginx     v16
         buildpack-null             v3
> dokku version: dokku version 0.28.1
> plugn version: plugn: 0.12.0build+3a27594
> dokku plugins: 
         00_dokku-standard    0.28.1 enabled    dokku core standard plugin
         20_events            0.28.1 enabled    dokku core events logging plugin
         app-json             0.28.1 enabled    dokku core app-json plugin
         apps                 0.28.1 enabled    dokku core apps plugin
         builder              0.28.1 enabled    dokku core builder plugin
         builder-dockerfile   0.28.1 enabled    dokku core builder-dockerfile plugin
         builder-herokuish    0.28.1 enabled    dokku core builder-herokuish plugin
         builder-lambda       0.27.0 enabled    dokku core builder-lambda plugin
         builder-null         0.28.1 enabled    dokku core builder-null plugin
         builder-pack         0.28.1 enabled    dokku core builder-pack plugin
         buildpacks           0.28.1 enabled    dokku core buildpacks plugin
         caddy-vhosts         0.28.1 enabled    dokku core caddy-vhosts plugin
         certs                0.28.1 enabled    dokku core certificate management plugin
         checks               0.28.1 enabled    dokku core checks plugin
         common               0.28.1 enabled    dokku core common plugin
         config               0.28.1 enabled    dokku core config plugin
         cron                 0.28.1 enabled    dokku core cron plugin
         docker-options       0.28.1 enabled    dokku core docker-options plugin
         domains              0.28.1 enabled    dokku core domains plugin
         enter                0.28.1 enabled    dokku core enter plugin
         git                  0.28.1 enabled    dokku core git plugin
         letsencrypt          0.18.1 enabled    Automated installation of let's encrypt TLS certificates
         logs                 0.28.1 enabled    dokku core logs plugin
         network              0.28.1 enabled    dokku core network plugin
         nginx-vhosts         0.28.1 enabled    dokku core nginx-vhosts plugin
         plugin               0.28.1 enabled    dokku core plugin plugin
         proxy                0.28.1 enabled    dokku core proxy plugin
         ps                   0.28.1 enabled    dokku core ps plugin
         registry             0.28.1 enabled    dokku core registry plugin
         repo                 0.28.1 enabled    dokku core repo plugin
         resource             0.28.1 enabled    dokku core resource plugin
         run                  0.28.1 enabled    dokku core run plugin
         scheduler            0.28.1 enabled    dokku core scheduler plugin
         scheduler-docker-local 0.28.1 enabled    dokku core scheduler-docker-local plugin
         scheduler-null       0.28.1 enabled    dokku core scheduler-null plugin
         shell                0.28.1 enabled    dokku core shell plugin
         ssh-keys             0.28.1 enabled    dokku core ssh-keys plugin
         storage              0.28.1 enabled    dokku core storage plugin
         trace                0.28.1 enabled    dokku core trace plugin
         traefik-vhosts       0.28.1 enabled    dokku core traefik-vhosts plugin
=====> painelgov-api app-json information
       App json computed selected:    app.json
       App json global selected:      app.json
       App json selected:             
=====> painelgov-api app information
       App created at:                1663537223
       App deploy source:             
       App deploy source metadata:    
       App dir:                       /home/dokku/painelgov-api
       App locked:                    false
=====> painelgov-api builder information
       Builder build dir:             
       Builder computed build dir:    
       Builder computed selected:     
       Builder global build dir:      
       Builder global selected:       
       Builder selected:              
=====> painelgov-api builder-dockerfile information
       Builder dockerfile computed dockerfile path: Dockerfile               
       Builder dockerfile global dockerfile path: Dockerfile               
       Builder dockerfile dockerfile path:                          
=====> painelgov-api builder-lambda information
       Builder lambda computed lambdayml path: lambda.yml               
       Builder lambda global lambdayml path: lambda.yml               
       Builder lambda lambdayml path:                          
=====> painelgov-api builder-pack information
       Builder pack computed projecttoml path: project.toml             
       Builder pack global projecttoml path: project.toml             
       Builder pack projecttoml path:                          
=====> painelgov-api buildpacks information
       Buildpacks computed stack:     gliderlabs/herokuish:latest-20
       Buildpacks global stack:       
       Buildpacks list:               
       Buildpacks stack:              
=====> painelgov-api ssl information
       Ssl dir:                       /home/dokku/painelgov-api/tls
       Ssl enabled:                   true                     
       Ssl hostnames:                 painelgov-api.mydomain.com.br
       Ssl expires at:                Dec 14 13:18:07 2022 GMT 
       Ssl issuer:                    C = US, O = Let's Encrypt, CN = R3
       Ssl starts at:                 Sep 15 13:18:08 2022 GMT 
       Ssl subject:                   subject=CN = painelgov-api.mydomain.com.br
       Ssl verified:                  self signed              
=====> painelgov-api checks information
       Checks disabled list:          none                     
       Checks skipped list:           none                     
=====> painelgov-api cron information
       Cron task count:               0
=====> painelgov-api docker options information
       Docker options build:                                   
       Docker options deploy:         --restart=on-failure:10  
       Docker options run:                                     
=====> painelgov-api domains information
       Domains app enabled:           true                     
       Domains app vhosts:            painelgov-api.mydomain.com.br
       Domains global enabled:        true                     
       Domains global vhosts:         mydomain.com.br             
=====> painelgov-api git information
       Git deploy branch:             master                   
       Git global deploy branch:      master                   
       Git keep git dir:              false                    
       Git rev env var:               GIT_REV                  
       Git sha:                                                
       Git last updated at:                                    
=====> painelgov-api letsencrypt information
       Letsencrypt active:            true                     
       Letsencrypt autorenew:         true                     
       Letsencrypt email:             myeeail@gmail.com       
       Letsencrypt expiration:        1671023887               
=====> painelgov-api logs information
       Logs computed max size:        10m
       Logs global max size:          10m
       Logs global vector sink:       
       Logs max size:                 
       Logs vector sink:              
=====> painelgov-api network information
       Network attach post create:           
       Network attach post deploy:           
       Network bind all interfaces:          false
       Network computed attach post create:  
       Network computed attach post deploy:  
       Network computed bind all interfaces: false
       Network computed initial network:     
       Network computed tld:                 
       Network global attach post create:    
       Network global attach post deploy:    
       Network global bind all interfaces:   false
       Network global initial network:       
       Network global tld:                   
       Network initial network:              
       Network static web listener:          
       Network tld:                          
       Network web listeners:                
=====> painelgov-api nginx information
       Nginx access log format:                                
       Nginx access log path:         /var/log/nginx/painelgov-api-access.log
       Nginx bind address ipv4:                                
       Nginx bind address ipv6:       ::                       
       Nginx client max body size:                             
       Nginx disable custom config:   false                    
       Nginx error log path:          /var/log/nginx/painelgov-api-error.log
       Nginx global hsts:             true                     
       Nginx computed hsts:           true                     
       Nginx hsts:                                             
       Nginx hsts include subdomains: true                     
       Nginx hsts max age:            15724800                 
       Nginx hsts preload:            false                    
       Nginx proxy buffer size:       4096                     
       Nginx proxy buffering:         on                       
       Nginx proxy buffers:           8 4096                   
       Nginx proxy busy buffers size: 8192                     
       Nginx proxy read timeout:      60s                      
       Nginx last visited at:         1663384787               
       Nginx x forwarded for value:   $remote_addr             
       Nginx x forwarded port value:  $server_port             
       Nginx x forwarded proto value: $scheme                  
       Nginx x forwarded ssl:                                  
=====> painelgov-api proxy information
       Proxy enabled:                 true
       Proxy port map:                http:80:5000 https:443:5000
       Proxy type:                    nginx
=====> painelgov-api ps information
       Deployed:                      false
       Processes:                     0
       Ps can scale:                  true
       Ps computed procfile path:     Procfile
       Ps global procfile path:       Procfile
       Ps procfile path:              
       Ps restart policy:             on-failure:10
       Restore:                       true
       Running:                       false
=====> painelgov-api registry information
       Registry computed image repo:      dokku/painelgov-api
       Registry computed push on release: false
       Registry computed server:          
       Registry global push on release:   
       Registry global server:            
       Registry image repo:               
       Registry push on release:          
       Registry server:                   
       Registry tag version:              
=====> painelgov-api resource information
=====> painelgov-api scheduler information
       Scheduler computed selected:   docker-local
       Scheduler global selected:     docker-local
       Scheduler selected:            
=====> painelgov-api scheduler-docker-local information
       Scheduler docker local disable chown:                          
       Scheduler docker local init process: true                     
       Scheduler docker local parallel schedule count:                          
=====> painelgov-api storage information

**Reply by josegonzalez** [reply]
This might be a memory issue - your server only has 1gb of memory. Nodejs builds tend to take much more memory. Adding swap memory may help.

**Reply by iober** [reply]
Are you sure? I use free heroku and it is only 512mb.

**Reply by josegonzalez** [reply]
Heroku builds do not have the same memory constraints as the runtime environment. I'm 95% confident this is related to memory pressure.

**Reply by iober** [reply]
Thanks @josegonzalez, u were right, I created a new VPS with 2GB of memory and it worked.
